### PR TITLE
Enforce let-punning in the janestreet profile

### PR DESCRIPTION
**Pull request by dvulakh**
Whenever it is possible to "pun" a `let`-binding, do so.

We do this for any `let` binding that is an extension or an operator and where the LHS and RHS are just the same identifier (no attached attributes, no type annotations).

I've added tests for a few tricky cases to `let_punning.ml`. I've also tested internally, and our styler runs without producing any syntax errors. More thorough testing will need to wait for the parser fix in compiler PR [#2612](https://github.com/ocaml-flambda/flambda-backend/pull/2612).

### CODE REVIEW

File: lib/Conf_t.ml
Comment (ccasin) [inline lib/Conf_t.ml]: `Always`?

File: lib/Fmt_ast.ml
Comment (ccasin) [inline lib/Fmt_ast.ml]: What is the `Location.contains` check here?  I would have guessed a pun was the only way for this to be true, and that it would always be true in that case.

**Review by ccasin — COMMENTED** [review]
I read this and it looks fine to me (modulo questions below).  But I'm going to hold off on approving until we've had a chance to test it on the tree because we could easily be missing some corner case.

File: lib/Fmt_ast.ml
Comment (dvulakh) [reply lib/Fmt_ast.ml]: I added a comment to explain what's going on here.

Comment (ccasin) [reply lib/Fmt_ast.ml]: Does the `lb_pun` in the original `Let_binding.t` tell us whether the original binding was punned?  If so, perhaps we can just not shadow it and check that rather than this locations math?

Comment (dvulakh) [reply lib/Fmt_ast.ml]: Good point. Fixed.

Comment (ccasin) [reply lib/Fmt_ast.ml]: Thanks!

**Review by ccasin — APPROVED** [review]
Approved, but will let David merge when he's ready.

**Pull request merged**
